### ISSUE TITLE
Rescheduling of a chain of Operations

### ISSUE THREAD

**Issue by gracinet** [AnyBlok/anyblok_wms_base#8]
in Anyblok / Wms Base, all Operation instances must have a date of execution (dt_execution field), even if they are merely planned. This is necessary so that peeking at stock levels in the future never counts the same Goods instance twice.

In many case, at the time of Operation creation, the issuer has of course no clue of the actual execution time, and if it has, chances are high that it'll turn out to be wrong anyway. This is not a problem, and it's explained in the documentation: the actual execution time will be taken into account upon execution. In other words, the functional meaning of that planned execution time can be anywhere betwen irrelevant to very important for the application.

However, currently, the dates get corrected at execution only for direct consequences of the current Operation. So, for instance, if we have two successive Avatars: `av1` and `av2`, respectively the outcome of `op1` and `op2`, an actual execution of `op1` later than `op2`'s scheduled dt_execution would push `av1`'s time range to overlap `av2`s (whereas an execution earlier than planned is expected never to be a problem).

In other words, we need a recursive rescheduling primitive, probably at the level of the Operation base class.
This issue to postpone actual implementation: this will be important at some point, but isn't urgent right now.

**Comment by petrus-v** [comment]
@gracinet 

I was reading [allow avatars with indeterminate time range](https://anyblok-wms-base.readthedocs.io/en/latest/improvements.html#allow-avatars-with-indeterminate-time-range) not sure if it should be a wms feature but could be a nice to have. You probably already through about it ?

In some businesses with chains operations we knows that the next operations can't be done before X time of the previous operations. For instance, (not sure if it's good example) if you have multiple warehouses and know that you needs 2 days to get there you cans says ``op2`` (receipt in warehouse 2) will be done in the best case 2 days after ``op1`` (departure from warehouse 1).

I believe that using relative time could give accurate projections, I don't think it's enough to only use relative time and I'm pretty sure that it won't avoid time computations on chained operations.  

my 2cts !

**Comment by gracinet** [comment]
@petrus-v, no, development hasn't started at all about [allow avatars with indeterminate time range](https://anyblok-wms-base.readthedocs.io/en/latest/improvements.html#allow-avatars-with-indeterminate-time-range).

About the remainings of your comment: yes lead time support will be a must-have, but I think it'll probably be in an optional Blok. That being said, something has already been done in preparation for that: [3392cf84a563d6a493469c3b08318480f3609250]  (will be released with 0.8) introduces a separate ``default_dt_execution`` classmethod which can be easily overridden by a dependent Blok.

About using relative times, I'm not sure what you really have in mind (a separate field, maybe). Care to elaborate about that ?

Thanks for the cents !

**Comment by gracinet** [comment]
For completeness, I should say that the classmethod can more importantly be overridden by a subclass, e.g., an applicative Blok may overriding Assembly to add an override of the classmethod defined by the base Operation class (or we could end up introducing lead times in the core, but I'm wary of bloat, since it'd need to expose a way for dependent Bloks to implement their lead time calculation logic anyway)

**Comment by petrus-v** [comment]
@gracinet 

Thanks for reply! indeed ``relative times`` is probably not the right term.

Not sure in English what's the difference between ``duration`` and ``lead times`` but that was my point to distinguish the expected amount of time for an operation and the expected date time when the operation id done (the current ``dt_execution`` field`as far I understand).

Thinking...  while planing something with some other constraints we just know that the end of the previous operation is not the beginning of next operation each operation should probably also maintains some constraints or queues.

For instance if:
* next operation ``op2`` can only be done between 8 o'clock  and 12 o'clock from Monday to Friday 
* next operation ``op2`` type is done by 1 and only one operator in a warehouse (a bottle neck) that can take only one operation per operation.

So if the previous operation ``op1`` ends on Saturday, the expected date time when the ``op2`` would be done is end of ``op1`` + first available schedule to manage ``op2`` type + duration to consume the ``op2`` type queue before that one + ``duration`` of ``op2`` type.

I definitively agree that needs to be split in different scheduled blok(s) to keep it robust in ``core``

**Comment by gracinet** [comment]
@petrus-v I was merely wondering if you were suggesting to store what would amount to ``datetime.timedelta`` on the avatars.
I'm not sure about the proper terminology to distinguish between the needed delay before the Operation is started, and the one between start and done, but it doens't matter much yet.

In the kind of complex scenario you're referring to, it's clear that a whole complex scheduling is needed, and it could involve more that dates (for instance reservation slots for machines etc.), and also, I've been thinking that it'll have to be asynchronous (i.e., handled by a different worker process or thread, in a different transaction): we don't want the operator flashing a late Arrival to pay the price (in waiting time) of computing all consequences.

Now, asynchronicity is a whole can of worms of its own – in particular because it's clear that Anyblok / WMS Base must not force to use a given type of inter worker communication. Even using PostgreSQL events (at least, PG itself is part of our assumptions) should still be changeable.

**Comment by gracinet** [comment]
version 0.9.0 will have a fix for this, that's probably going to be suboptimal in terms of performance, but can be later improved when we add Avatar indeterminate dates.

For now it's already a very good thing to be able to enforce `dt_until >= dt_from`, especially with production starting now
